### PR TITLE
ECL-668: Amend Registration - Feature Flag

### PR DESCRIPTION
**Pull request by nemanjatomanovic**

### CODE REVIEW

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala [hunk 1]
```diff
@@ -61,6 +70,22 @@ class AmendRegistrationStartController @Inject() (
...
 
+  private def routeToAmendReason(eclRegistrationReference: String, internalId: String)(implicit
+    hc: HeaderCarrier
+  ): Future[Result] = {
+    val result = for {
+      registration            <- registrationService.getOrCreateRegistration(internalId)
+      getSubscriptionResponse <- registrationService.getSubscription(eclRegistrationReference)
+      upsertedRegistration    <-
+        registrationService.upsertRegistration(
+          registrationService.transformToRegistration(registration, getSubscriptionResponse)
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala:81]: can the result of registrationService.transformToRegistration(registration, getSubscriptionResponse) be set to a val in the for-comprehension? It's easier to process than a call within a call:

```
registration = registrationService.transformToRegistration(registration, getSubscriptionResponse)
_ <-   registrationService.upsertRegistration(registration)
```

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala [hunk 2]
```diff
@@ -40,7 +42,8 @@ class AmendRegistrationStartController @Inject() (
   errorHandler: ErrorHandler,
   view: AmendRegistrationStartView,
   registrationService: EclRegistrationService,
-  registrationConnector: EclRegistrationConnector
+  registrationConnector: EclRegistrationConnector,
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala:45]: why are we using the connector here rather than registrationService?

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala [hunk 3]
```diff
@@ -61,6 +70,22 @@ class AmendRegistrationStartController @Inject() (
...
+    val result = for {
+      registration            <- registrationService.getOrCreateRegistration(internalId)
+      getSubscriptionResponse <- registrationService.getSubscription(eclRegistrationReference)
+      upsertedRegistration    <-
+        registrationService.upsertRegistration(
+          registrationService.transformToRegistration(registration, getSubscriptionResponse)
+        )
+    } yield upsertedRegistration
+
+    result.map(_ => Redirect(routes.AmendReasonController.onPageLoad(CheckMode).url))
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala:85]: we're not handling when an error happens

File: app/uk/gov/hmrc/economiccrimelevyregistration/models/BusinessSector.scala [hunk 4]
```diff
@@ -83,4 +93,20 @@ object BusinessSector {
 
     override def writes(o: BusinessSector): JsValue = JsString(o.toString)
   }
+
+  def transformFromSubscriptionResponse(businessSector: String): BusinessSector =
+    businessSector.replace(" ", "").toLowerCase() match {
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/models/BusinessSector.scala:98]: you can use `businessSector.filterNot(_.isWhitespace).toLowerCase`

File: app/uk/gov/hmrc/economiccrimelevyregistration/models/BusinessSector.scala [hunk 5]
```diff
@@ -83,4 +93,20 @@ object BusinessSector {
...
+      case "auditor"                       => Auditor
+      case "insolvencypractitioner"        => InsolvencyPractitioner
+      case "externalaccountant"            => ExternalAccountant
+      case "taxadviser"                    => TaxAdviser
+      case "independentlegalprofessional"  => IndependentLegalProfessional
+      case "trustorcompanyserviceprovider" => TrustOrCompanyServiceProvider
+      case "estateagentorlettingagent"     => EstateAgentOrLettingAgent
+      case "highvaluedealer"               => HighValueDealer
+      case "cryptoassetexchangeprovider"   => CryptoAssetExchangeProvider
+      case s                               => throw new IllegalStateException(s"$s is not a valid BusinessSector")
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/models/BusinessSector.scala:110]: :( but I can refactor this

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala [hunk 6]
```diff
@@ -48,4 +50,79 @@ class EclRegistrationService @Inject() (
...
+      case Some(value) =>
+        ContactDetails(
+          Some(value.name),
+          Some(value.positionInCompany),
+          Some(value.emailAddress),
+          Some(value.telephone)
+        )
+      case _           => ContactDetails.empty
+    }
+    val secondContactPresent                = if (secondaryContact.isEmpty) {
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala:81]: this doesn't have to be wrapped in an if block

`val secondContactPresent = Some(secondaryContact.isDefined)`

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala [hunk 7]
```diff
@@ -48,4 +50,79 @@ class EclRegistrationService @Inject() (
...
+    val address: EclAddress            = EclAddress(
+      None,
+      Some(subscriptionAddress.addressLine1),
+      subscriptionAddress.addressLine2,
+      subscriptionAddress.addressLine3,
+      subscriptionAddress.addressLine4,
+      None,
+      subscriptionAddress.postCode,
+      None,
+      subscriptionAddress.countryCode.getOrElse(
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala:100]: getOrElse exception?

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala [hunk 8]
```diff
@@ -48,4 +50,79 @@ class EclRegistrationService @Inject() (
...
+      businessSector = Some(businessSector),
+      contactAddress = Some(address),
+      partnershipName = getSubscriptionResponse.legalEntityDetails.organisationName,
+      amlSupervisor = Some(getAmlSupervisor(getSubscriptionResponse.additionalDetails.amlSupervisor))
+    )
+  }
+
+  private def getAmlSupervisor(amlSupervisor: String): AmlSupervisor = {
+    val amlProfessionalBodySupervisors: Option[String] =
+      appConfig.amlProfessionalBodySupervisors.find(p =>
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala:115]:
```
val updatedAmlSupervisor = amlSupervisor.filterNot(_.isWhitespace).toLowerCase

appConfig.amlProfessionalBodySupervisors.exists(updatedAmlSupervisor)
```

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala [hunk 9]
```diff
@@ -48,4 +50,79 @@ class EclRegistrationService @Inject() (
...
+        p.toLowerCase() == amlSupervisor.replace(" ", "").toLowerCase()
+      )
+
+    if (amlProfessionalBodySupervisors.isEmpty) {
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala:119]: if we need to strip spaces from amlSupervisor twice and lowercase, we should save that in a val

File: test/uk/gov/hmrc/economiccrimelevyregistration/connectors/EclRegistrationConnectorSpec.scala [hunk 10]
```diff
@@ -292,4 +294,27 @@ class EclRegistrationConnectorSpec extends SpecBase {
         reset(mockHttpClient)
     }
   }
+
+  "getSubscription" should {
+    "return a subscription response with answers that user already provided during the registration" in forAll(
```
Comment (liandra-giacomine) [inline test/uk/gov/hmrc/economiccrimelevyregistration/connectors/EclRegistrationConnectorSpec.scala:299]: need a test for a negative scenario

File: test/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartControllerSpec.scala [hunk 11]
```diff
@@ -45,30 +45,37 @@ class AmendRegistrationStartControllerSpec extends SpecBase {
...
-    "return OK and the correct view" in forAll { registration: Registration =>
-      when(
-        mockRegistrationAdditionalInfoService.createOrUpdate(
-          anyString(),
-          any()
-        )(any())
-      ).thenReturn(
-        Future.successful(())
-      )
+    "return OK and the correct view" in forAll {
```
Comment (liandra-giacomine) [inline test/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartControllerSpec.scala:53]: need a test for a negative scenario

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala (hunk 1 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala]: Done

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala (hunk 2 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala]: Done

File: app/uk/gov/hmrc/economiccrimelevyregistration/models/BusinessSector.scala (hunk 4 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/models/BusinessSector.scala]: Done

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala (hunk 3 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartController.scala]: Done

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala (hunk 6 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala]: Done

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala (hunk 7 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala]: Done

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala (hunk 9 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationService.scala:112]: Done

File: test/uk/gov/hmrc/economiccrimelevyregistration/connectors/EclRegistrationConnectorSpec.scala (hunk 10 above)
Comment (nemanjatomanovic) [reply test/uk/gov/hmrc/economiccrimelevyregistration/connectors/EclRegistrationConnectorSpec.scala:299]: Done :)

File: test/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartControllerSpec.scala (hunk 11 above)
Comment (nemanjatomanovic) [reply test/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmendRegistrationStartControllerSpec.scala:50]: Done :)

File: test/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationServiceSpec.scala [hunk 12]
```diff
@@ -55,4 +57,18 @@ class EclRegistrationServiceSpec extends SpecBase {
       result shouldBe registration
     }
   }
+
+  "getSubscription" should {
+    "return valid subscription response when eclReference is passed" in forAll(
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply test/uk/gov/hmrc/economiccrimelevyregistration/services/EclRegistrationServiceSpec.scala:66]: Done :)

**Review by liandra-giacomine — APPROVED** [review]
